decompiler/pipeline/controlflowanalysis/variable_name_generation.py

```diff
@@ -0,0 +1,123 @@
+    """Returns a list of variables contained in this dataflow object"""
+    variables: List[Variable] = []
+    for sub_exp in dfo.subexpressions():
+        if isinstance(sub_exp, Variable):
+            variables.append(sub_exp)
+    return variables
+
+
+class NamingConvention(str, Enum):
+    default = "default"
+    system_hungarian = "system_hungarian"
+
+
+class VariableNameGeneration(PipelineStage):
+    """ 
+    Pipelinestage in charge of renaming variables to a configured format.
+    Currently only the 'default' or 'hungarian' system are supported.
+    """
+
+    name : str = "variable-name-generation"
+    type_prefix = {
+        Float: {16: "h", 32: "f", 64: "d", 80: "ld", 128: "q", 256: "o"},
+        Integer: {8: "ch", 16: "s", 32: "i", 64: "l", 128: "i128"},
+    }
+
+
+    def __init__(self):
+        self._ast: Optional[AbstractSyntaxTree] = None
+        self._notation: Optional[str] = None
+        self._var_name: Optional[str] = None
+        self._pointer_base: bool = True
+        self._type_separator: str = ""
+        self._counter_separator: str = ""
+        self._variables: List[Variable] = []
+        self._params : List[Variable] = []
+
+
+    def run(self, task: DecompilerTask):
+        self._ast = task.syntax_tree
+        self._notation = task.options.getstring(f"{self.name}.notation", fallback="default")
+        self._var_name = task.options.getstring(f"{self.name}.variable_name", fallback="Var")
+        self._pointer_base = task.options.getboolean(f"{self.name}.pointer_base", fallback=True)
+        self._type_separator = task.options.getstring(f"{self.name}.type_separator", fallback="")
+        self._counter_separator = task.options.getstring(f"{self.name}.counter_separator", fallback="")
+        self._params = task.function_parameters
+
+        if self._notation != NamingConvention.default:
+            self._collect()
+            self._purge()
+            self._rename()
+
+
+    def _collect(self):
+        """Collects all variables by iterating over every node in the AST."""
+        for node in self._ast.topological_order():
+            if isinstance(node, CodeNode):
+                for stmt in node.instructions:
+                    self._variables.extend(_get_containing_variables(stmt))
+            elif isinstance(node, (ConditionNode, LoopNode)):
+                for expr in [self._ast.condition_map[symbol] for symbol in node.condition.get_symbols()]:
+                    self._variables.extend(_get_containing_variables(expr))
+            elif isinstance(node, (SwitchNode, CaseNode)):
+                self._variables.extend(_get_containing_variables(node.expression))         
+
+
+    def _purge(self):
+        """Remove variables for renaming if they are:
+            - function params 
+            - loop variables which have been renamed by For/WhileLoopRenamer
+            - global vars with a symbol as a name
+        """
```

Comment:
Kinda of weird to have this in another function, since it allows the class to be in an illegal state when only one of the functions is invoked.

Reply:
Merged the variable collection and purging together. 
Now after the method `collectVariableNames` is called, every variable which should be renamed is in a list.